src/relay/ir/debug_printer.cc

```diff
@@ -0,0 +1,305 @@
+    }
+    return vec;
+  }
+
+  std::vector<Doc> DocifyParamArray(const tvm::Array<Param>& arr) {
+    std::vector<Doc> vec;
+    for (size_t i = 0; i < arr.size(); ++i) {
+      vec.push_back(Docify(arr[i]));
+    }
+    return vec;
+  }
+
+  Doc VisitExpr_(const ConstantNode* c) final {
+    return DocOfStr("some_constant");
+  }
+
+  Doc VisitExpr_(const TupleNode* t) final {
+    return Seq("<", DocifyExprArray(t->fields), ">");
+  }
+
+  Doc VisitExpr_(const VarNode* v) final {
+    return DocOfStr(VarName(GetRef<Var>(v)));
+  }
+
+  Doc VisitExpr_(const GlobalVarNode* g) final {
+    return DocOfStr(g->name_hint);
+  }
+
+  Doc VisitExpr_(const ParamNode* p) final {
+    return TypeAnnotation(Docify(p->var), p->type);
+  }
+
+  Doc VisitExpr_(const FunctionNode* f) final {
+    return Group(TypeAnnotation(Seq("(", DocifyParamArray(f->params), ")"), f->ret_type) + Sep() +
+                 DocOfStr("=>") + Sep() +
+                 Block(indent, "{", Docify(f->body), "}"));
+  }
+
+  Doc VisitExpr_(const CallNode* c) final {
+    auto args = DocifyExprArray(c->args);
+    return Docify(c->op) + Seq("<", DocifyExprArray(c->args), ">");
+  }
+
+  Doc VisitExpr_(const LetNode* l) final {
+    return Group(DocOfStr("let") + Sep() + TypeAnnotation(Docify(l->var), l->value_type) + Sep() +
+                 DocOfStr("=") + Sep() + Docify(l->value) + DocOfStr(";") + Endl() +
+                 Docify(l->body));
+  }
+
+  Doc VisitExpr_(const IfNode* i) final {
+    return Group(DocOfStr("if") + Sep() + Docify(i->cond) + Sep() +
+                 Block(indent, "{", Docify(i->true_branch), "}") + Sep() +
+                 DocOfStr("else") + Sep() +
+                 Block(indent, "{", Docify(i->false_branch), "}"));
+  }
+
+  Doc VisitExpr_(const OpNode* o) final {
+    return DocOfStr(o->name);
+  }
+
+ public:
+  ExprDocifier(const Environment& env) : env(env), td(env) { }
+
+  Doc Docify(const Expr& e) { return (*this)(e); }
+};
+
+Doc DocOfExpr(const Environment& env, const Expr& expr) {
+  ExprDocifier d(env);
+  return d.Docify(expr);
+}
+
+Doc DocOfType(const Environment& env, const Type& expr) {
+  TypeDocifier d(env);
+  return d.Docify(expr);
+}
+
+RDoc ExprRDoc(const Environment& env, const Expr& expr) {
+  return Layout(DocOfExpr(env, expr));
+}
+
+RDoc TypeRDoc(const Environment& env, const Type& expr) {
+  return Layout(DocOfType(env, expr));
+}
+
+std::ostream & DebugPrint(const Environment& env, const Expr& e, std::ostream& os) {
+  return os << ExprRDoc(env, e);
+}
+
+std::ostream & DebugPrint(const Environment& env, const Type& t, std::ostream& os) {
+  return os << TypeRDoc(env, t);
+}
+
+std::string PrintExpr(const Environment& env, const Expr& e) {
+  std::stringstream ss;
+  ss << ExprRDoc(env, e);
+  return ss.str();
+}
+
+std::string PrintType(const Environment& env, const Type& t) {
+  std::stringstream ss;
+  ss << TypeRDoc(env, t);
+  return ss.str();
+}
+
+TVM_REGISTER_API("relay._expr._debug_print_expr")
```

Comment:
We can have a single debug_print which checks if a NodeRef is Expr or Type, then call specific ones.

Reply:
how do we check it in C++? xxx = NodeRef, xxx.as ? ... : ... ?

Reply:
NodeRef x = args[1];

if (x.as<TypeNode>()) {
   Type t = Downcast<Type>(x);
} else { 
   ....
}

Reply:
done